## Introduction 


Name : Kevin Giovanni Pradana

Data Attributes 

1. event_time : when the event happen
2. event_type : types of event
3. product_id : id for each product 
4. category_id : id for product category
5. brand : brand name of product
6. price : product price
7. user_id : who is buying the product
8. user_session : user session purchase

### Libraries

In [5]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
import sys

### Data Loading

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [7]:
#loading data
df = pd.read_csv('/content/drive/MyDrive/data_hackhaton/first_question/user_events.csv')

In [8]:
df.head()

event_time event_type  product_id          category_id  \
0  2020-09-24 11:57:06 UTC       view     1996170  2144415922528452715   
1  2020-09-24 11:57:26 UTC       view      139905  2144415926932472027   
2  2020-09-24 11:57:27 UTC       view      215454  2144415927158964449   
3  2020-09-24 11:57:33 UTC       view      635807  2144415923107266682   
4  2020-09-24 11:57:36 UTC       view     3658723  2144415921169498184   

                   category_code        brand   price              user_id  \
0          electronics.telephone          NaN   31.90  1515915625519388267   
1    computers.components.cooler       zalman   17.16  1515915625519380411   
2                            NaN          NaN    9.81  1515915625513238515   
3  computers.peripherals.printer       pantum  113.81  1515915625519014356   
4                            NaN  cameronsino   15.87  1515915625510743344   

  user_session  
0   LJuJVLEjPT  
1   tdicluNnRY  
2   4TMArHtXQy  
3   aGFYrNgC08  
4   aa4mmk0kwQ

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 885129 entries, 0 to 885128
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   event_time     885129 non-null  object 
 1   event_type     885129 non-null  object 
 2   product_id     885129 non-null  int64  
 3   category_id    885129 non-null  int64  
 4   category_code  648910 non-null  object 
 5   brand          672765 non-null  object 
 6   price          885129 non-null  float64
 7   user_id        885129 non-null  int64  
 8   user_session   884964 non-null  object 
dtypes: float64(1), int64(3), object(5)
memory usage: 60.8+ MB


there are 885129 rows of data, but event_time data type is not particularly correct, that's why we need to change it to **datetime** and there are missing values in category_code and brand columns

### Data Exploration

In [10]:
df.tail()

event_time event_type  product_id          category_id  \
885124  2021-02-28 23:55:01 UTC       view      953226  2144415927553229037   
885125  2021-02-28 23:58:05 UTC       view     1715907  2144415927049912542   
885126  2021-02-28 23:58:09 UTC       view     4170534  2144415939364389423   
885127  2021-02-28 23:58:14 UTC       view      888273  2144415921932861531   
885128  2021-02-28 23:59:09 UTC       view      743182  2144415935631458761   

                       category_code     brand   price              user_id  \
885124                           NaN       NaN  219.94  1515915625611023730   
885125          electronics.video.tv  starwind   80.03  1515915625611024014   
885126            electronics.clocks   amazfit   64.92  1515915625611024020   
885127         electronics.telephone       NaN   10.16  1515915625611024030   
885128  construction.tools.soldering      kada   65.08  1515915625556087775   

       user_session  
885124   FRLqIttxKU  
885125   g6WqPf50Ma  
885126   xNIJBqZdkd  
885127   9pCbKMIcSx  
885128   BejOXRngEW

In [11]:
df.head()

event_time event_type  product_id          category_id  \
0  2020-09-24 11:57:06 UTC       view     1996170  2144415922528452715   
1  2020-09-24 11:57:26 UTC       view      139905  2144415926932472027   
2  2020-09-24 11:57:27 UTC       view      215454  2144415927158964449   
3  2020-09-24 11:57:33 UTC       view      635807  2144415923107266682   
4  2020-09-24 11:57:36 UTC       view     3658723  2144415921169498184   

                   category_code        brand   price              user_id  \
0          electronics.telephone          NaN   31.90  1515915625519388267   
1    computers.components.cooler       zalman   17.16  1515915625519380411   
2                            NaN          NaN    9.81  1515915625513238515   
3  computers.peripherals.printer       pantum  113.81  1515915625519014356   
4                            NaN  cameronsino   15.87  1515915625510743344   

  user_session  
0   LJuJVLEjPT  
1   tdicluNnRY  
2   4TMArHtXQy  
3   aGFYrNgC08  
4   aa4mmk0kwQ

In [12]:
df.select_dtypes("number").describe()

product_id   category_id          price       user_id
count  8.851290e+05  8.851290e+05  885129.000000  8.851290e+05
mean   1.906621e+06  2.144423e+18     146.328713  1.515916e+18
std    1.458708e+06  6.165105e+14     296.807683  3.554165e+07
min    1.020000e+02  2.144416e+18       0.220000  1.515916e+18
25%    6.988030e+05  2.144416e+18      26.460000  1.515916e+18
50%    1.452883e+06  2.144416e+18      65.710000  1.515916e+18
75%    3.721194e+06  2.144416e+18     190.490000  1.515916e+18
max    4.183880e+06  2.227847e+18   64771.060000  1.515916e+18

In [13]:
df.select_dtypes("object").describe().T

count  unique                              top    freq
event_time     885129  845041          2021-02-04 21:48:32 UTC      18
event_type     885129       3                             view  793748
category_code  648910     107  computers.components.videocards  116717
brand          672765     999                             asus   27706
user_session   884964  490398                       nFlhu5QzOd     572

we can see that the top product brand is ASUS and mostly people buy videocards

In [14]:
df["category_code"].value_counts()

computers.components.videocards     116717
electronics.telephone                84360
computers.peripherals.printer        43224
stationery.cartrige                  38725
electronics.audio.acoustic           26766
                                     ...  
furniture.bedroom.blanket               26
appliances.kitchen.refrigerators        15
furniture.kitchen.chair                 15
accessories.briefcase                   13
auto.accessories.winch                   6
Name: category_code, Length: 107, dtype: int64

count products purchases based on general category

In [15]:
df["category_code_general"] = df["category_code"].str.split(".", expand=True)[0].astype('category')

In [16]:
print(df["category_code_general"].nunique())

14


In [17]:
df["category_code_general"].value_counts()

computers       316954
electronics     171108
stationery       42984
appliances       41087
auto             35433
construction     31067
furniture         3365
country_yard      3142
accessories       2080
medicine           713
kids               372
jewelry            255
sport              232
apparel            118
Name: category_code_general, dtype: int64

### Data Cleaning

In [18]:
data = df.copy()

In [19]:
null_columns = data.columns[data.isnull().any()]

In [20]:
data[null_columns].isnull().sum()

category_code            236219
brand                    212364
user_session                165
category_code_general    236219
dtype: int64

In [21]:
data.isnull().sum()

event_time                    0
event_type                    0
product_id                    0
category_id                   0
category_code            236219
brand                    212364
price                         0
user_id                       0
user_session                165
category_code_general    236219
dtype: int64

In [22]:
data['event_time']=pd.to_datetime(data['event_time'])

In [23]:
data["category_code"].value_counts()

computers.components.videocards     116717
electronics.telephone                84360
computers.peripherals.printer        43224
stationery.cartrige                  38725
electronics.audio.acoustic           26766
                                     ...  
furniture.bedroom.blanket               26
appliances.kitchen.refrigerators        15
furniture.kitchen.chair                 15
accessories.briefcase                   13
auto.accessories.winch                   6
Name: category_code, Length: 107, dtype: int64

In [24]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 885129 entries, 0 to 885128
Data columns (total 10 columns):
 #   Column                 Non-Null Count   Dtype              
---  ------                 --------------   -----              
 0   event_time             885129 non-null  datetime64[ns, UTC]
 1   event_type             885129 non-null  object             
 2   product_id             885129 non-null  int64              
 3   category_id            885129 non-null  int64              
 4   category_code          648910 non-null  object             
 5   brand                  672765 non-null  object             
 6   price                  885129 non-null  float64            
 7   user_id                885129 non-null  int64              
 8   user_session           884964 non-null  object             
 9   category_code_general  648910 non-null  category           
dtypes: category(1), datetime64[ns, UTC](1), float64(1), int64(3), object(4)
memory usage: 61.6+ MB


In [25]:
data.dropna(subset=['category_code'],inplace=True)

In [26]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 648910 entries, 0 to 885128
Data columns (total 10 columns):
 #   Column                 Non-Null Count   Dtype              
---  ------                 --------------   -----              
 0   event_time             648910 non-null  datetime64[ns, UTC]
 1   event_type             648910 non-null  object             
 2   product_id             648910 non-null  int64              
 3   category_id            648910 non-null  int64              
 4   category_code          648910 non-null  object             
 5   brand                  516349 non-null  object             
 6   price                  648910 non-null  float64            
 7   user_id                648910 non-null  int64              
 8   user_session           648792 non-null  object             
 9   category_code_general  648910 non-null  category           
dtypes: category(1), datetime64[ns, UTC](1), float64(1), int64(3), object(4)
memory usage: 50.1+ MB


In [27]:
data[null_columns].isnull().sum()

category_code                 0
brand                    132561
user_session                118
category_code_general         0
dtype: int64

In [28]:
data.head()

event_time event_type  product_id          category_id  \
0 2020-09-24 11:57:06+00:00       view     1996170  2144415922528452715   
1 2020-09-24 11:57:26+00:00       view      139905  2144415926932472027   
3 2020-09-24 11:57:33+00:00       view      635807  2144415923107266682   
5 2020-09-24 11:57:59+00:00       view      664325  2144415951611757447   
6 2020-09-24 11:58:23+00:00       view     3791349  2144415935086199225   

                   category_code   brand   price              user_id  \
0          electronics.telephone     NaN   31.90  1515915625519388267   
1    computers.components.cooler  zalman   17.16  1515915625519380411   
3  computers.peripherals.printer  pantum  113.81  1515915625519014356   
5         construction.tools.saw  carver   52.33  1515915625519388062   
6              computers.desktop     NaN  215.41  1515915625519388877   

  user_session category_code_general  
0   LJuJVLEjPT           electronics  
1   tdicluNnRY             computers  
3   aGFYrNgC08             computers  
5   vnkdP81DDW          construction  
6   J1t6sIYXiV             computers

In [29]:
data.dropna(subset=['user_session'],inplace=True)

In [30]:
data[null_columns].isnull().sum()

category_code                 0
brand                    132536
user_session                  0
category_code_general         0
dtype: int64

In [31]:
data.dropna(subset=['user_session'],inplace=True)

### Exploratory Data Analysis 

In [32]:
import plotly.express as px
import plotly.graph_objs as go

In [33]:
# best performing brands

best_perform_brand = data.groupby('brand')['price'].sum().reset_index().sort_values('price', ascending=False).head(10)

In [34]:
fig = px.bar(
    best_perform_brand, 
    x='brand', 
    y='price', 
    title='Best performing brands',
    width=800, 
    height=800
)

fig.show()

In [35]:
#most sold brands
most_sold_brands = data.groupby('brand')['price'].agg('count').reset_index().sort_values('price', ascending=False).head(10)
most_sold_brands.rename(columns={"brand": "brand", "price": "times_sold"}, inplace=True)

In [36]:
fig = px.bar(
    most_sold_brands, 
    x='brand', 
    y='times_sold', 
    title='Most sold brands',
    width=800, 
    height=800
)

fig.show()

In [37]:
#most bought categories
most_sold_categories = data.groupby('category_code_general')['price'].agg('count').reset_index().sort_values('price', ascending=False).head(10)
most_sold_categories.rename(columns={"category_code_general": "category", "price": "times_sold"}, inplace=True)

In [38]:
fig = px.bar(
    most_sold_categories, 
    x='category', 
    y='times_sold', 
    title='Most sold categories',
    width=1500, 
    height=800
)

fig.show()

In [39]:
best_month = data.groupby(data['event_time'].dt.strftime('%B'))['price'].sum().sort_values()

In [40]:
fig = px.bar(
    best_month,
    x="price", 
    orientation='h', 
    title="In which months most of the purchases were made", 
    width=800, 
    height=800
)

fig.show()

In [41]:
#best time for purchase. filter with lambda <900 as a lot of data was on 1970-1-1 12 am
best_time = data.groupby(data['event_time'].dt.strftime('%r'))['price'].sum().sort_values()[lambda x: x<= 900000].tail(10)

In [42]:
fig = px.bar(
    best_time,
    x="price", 
    orientation='h', 
    title="At what time most of the purchases were made", 
    width=800, 
    height=800
)

fig.show()

In [44]:
data.to_csv('clean_data.csv', index = False, encoding='utf-8')

In [45]:
data.head()

event_time event_type  product_id          category_id  \
0 2020-09-24 11:57:06+00:00       view     1996170  2144415922528452715   
1 2020-09-24 11:57:26+00:00       view      139905  2144415926932472027   
3 2020-09-24 11:57:33+00:00       view      635807  2144415923107266682   
5 2020-09-24 11:57:59+00:00       view      664325  2144415951611757447   
6 2020-09-24 11:58:23+00:00       view     3791349  2144415935086199225   

                   category_code   brand   price              user_id  \
0          electronics.telephone     NaN   31.90  1515915625519388267   
1    computers.components.cooler  zalman   17.16  1515915625519380411   
3  computers.peripherals.printer  pantum  113.81  1515915625519014356   
5         construction.tools.saw  carver   52.33  1515915625519388062   
6              computers.desktop     NaN  215.41  1515915625519388877   

  user_session category_code_general  
0   LJuJVLEjPT           electronics  
1   tdicluNnRY             computers  
3   aGFYrNgC08             computers  
5   vnkdP81DDW          construction  
6   J1t6sIYXiV             computers

In [47]:
from sklearn.cluster import KMeans

### Data Preprocessing